framework/src/modules/chain/blocks/chain.js

```diff
@@ -0,0 +1,439 @@
+	await stateStore.round.finalize();
+};
+
+/**
+ * Calls saveBlock for the block and performs round tick
+ *
+ * @private
+ * @param {Object} block - Block object
+ * @param {boolean} saveBlock - Flag to save block into database
+ * @param {function} tx - Database transaction
+ * @returns {Promise<reject|resolve>}
+ */
+const saveBlockStep = async (storage, roundsModule, block, shouldSave, tx) => {
+	if (shouldSave) {
+		await saveBlock(storage, block, tx);
+	}
+	await new Promise((resolve, reject) => {
+		roundsModule.tick(
+			block,
+			tickErr => {
+				if (tickErr) {
+					return reject(tickErr);
+				}
+				return resolve();
+			},
+			tx
+		);
+	});
+};
+
+/**
+ * Reverts confirmed transactions due to block deletion
+ * @param {Object} block - secondLastBlock
+ * @param {Object} tx - database transaction
+ */
+const undoConfirmedStep = async (storage, slots, block, exceptions, tx) => {
+	if (block.transactions.length === 0) {
+		return;
+	}
+
+	const nonInertTransactions = block.transactions.filter(
+		transaction => !exceptions.inertTransactions.includes(transaction.id)
+	);
+
+	const { stateStore, transactionsResponses } = await undoTransactions(
+		storage,
+		exceptions
+	)(nonInertTransactions, tx);
+
+	const unappliedTransactionResponse = transactionsResponses.find(
+		transactionResponse => transactionResponse.status !== TransactionStatus.OK
+	);
+
+	if (unappliedTransactionResponse) {
+		throw unappliedTransactionResponse.errors;
+	}
+
+	await stateStore.account.finalize();
+
+	stateStore.round.setRoundForData(slots.calcRound(block.height));
+
+	await stateStore.round.finalize();
+};
+
+/**
+ * Performs backward tick
+ * @param {Object} oldLastBlock - secondLastBlock
+ * @param {Object} previousBlock - block to delete
+ * @param {Object} tx - database transaction
+ */
+const backwardTickStep = async (
+	roundsModule,
+	oldLastBlock,
+	previousBlock,
+	tx
+) =>
+	new Promise((resolve, reject) => {
+		// Perform backward tick on rounds
+		// WARNING: DB_WRITE
+		roundsModule.backwardTick(
+			oldLastBlock,
+			previousBlock,
+			backwardTickErr => {
+				if (backwardTickErr) {
+					return reject(backwardTickErr);
+				}
+				return resolve();
+			},
+			tx
+		);
+	});
+
+/**
+ * Deletes last block, undo transactions, recalculate round.
+ *
+ * @param  {function} cb - Callback function
+ * @returns {function} cb - Callback function from params (through setImmediate)
+ * @returns {Object} cb.err - Error
+ * @returns {Object} cb.obj - New last block
+ */
+const popLastBlock = async (
+	storage,
+	interfaceAdapters,
+	genesisBlock,
+	roundsModule,
+	slots,
+	oldLastBlock
+) => {
+	let secondLastBlock;
+	await storage.entities.Block.begin('Chain:deleteBlock', async tx => {
+		secondLastBlock = await loadSecondLastBlock(
+			storage,
+			interfaceAdapters,
+			genesisBlock,
+			oldLastBlock.previousBlock,
+			tx
+		);
+		await undoConfirmedStep(storage, slots, oldLastBlock, tx);
```

Comment:
This function expects `storage, slots, block, exceptions, tx` as params but `exceptions` is missing here